_posts/2017-03-05-artsy-technology-stack-2017.md

```diff
@@ -0,0 +1,169 @@
+
+For more in-depth work, we use [Jupyter Notebooks](https://ipython.org/notebook.html) to connect to our Redshift cluster and by default import [pandas](http://pandas.pydata.org), [sci-kit learn](http://scikit-learn.org/stable), and [pyplot](http://matplotlib.org/api/pyplot_api.html) for data analysis.
+
+## Search
+
+We completed our full migration from [Solr](http://lucene.apache.org/solr) to [Elasticsearch](https://www.elastic.co) in the last 18 months, and now use the latter across artsy.net, from all our artwork filter interfaces through to our real-time artwork similarity feature. Elasticsearch gives us high availability clustering features out of the box and easy horizontal scaling.
+
+## Platform Services
+
+As Artsy's business has grown more complex, so has the data and concepts handled by its core API. We've begun supporting certain product areas with separate, dedicated API services, and even extracting existing API domains into separate services when practical. These services tend to expose simple [REST](https://en.wikipedia.org/wiki/Representational_state_transfer)-ful HTTP APIs, maintain separate data sources, and even do their own [authentication](/blog/2016/10/26/jwt-artsy-journey). This has certain advantages:
+
+* Each system can be deployed and scaled independently.
+* Each chooses the best-suited languages and technologies for its purpose.
+* Code bases remain more focused and developers' cognitive overhead is minimized.
+
+Balancing these out are some very real disadvantages:
+
+* Development must sometimes touch multiple systems.
+* Some data is copied between services. These can become out-of-sync, though we always try to have a single _authoritative_ source in such cases.
+* Deploys must be coordinated.
+
+At our size and complexity, a single code base is simply impractical. So, we've tried to be consistent in the coding, deployment, monitoring, and logging practices of these services. The more repeatable and disciplined our process, the less overhead is introduced by additional systems.
+
+We've also explored alternate communication patterns, so systems aren't as dependent on each other's APIs. Recently we've begun publishing a stream of interesting data events from our core systems. Other systems can simply subscribe to the notifications they care about, so the source system doesn't need to be concerned about integrating with one more destination. After experimenting with [Kafka](https://kafka.apache.org) but finding it hard to manage, we switched to [RabbitMQ](https://www.rabbitmq.com) for this purpose. To provide consistency when publishing events we have [our own gem](https://github.com/artsy/artsy-eventservice).
```

Comment:
I find `interesting events` to be strange here. Interesting in which sense? Maybe something like `Recently we've begun publishing a stream of data events from our core systems that other systems can consume.`